torch/_overrides.py

```diff
@@ -0,0 +1,264 @@
+
+    Arguments
+    ---------
+    implementation : function
+        Function that implements the operation on ``torch.Tensor`` without
+        overrides when called like ``implementation(*args, **kwargs)``.
+    public_api : function
+        Function exposed by the public torch API originally called like
+        ``public_api(*args, **kwargs)`` on which arguments are now being
+        checked.
+    relevant_args : iterable
+        Iterable of arguments to check for __torch_function__ methods.
+    args : tuple
+        Arbitrary positional arguments originally passed into ``public_api``.
+    kwargs : tuple
+        Arbitrary keyword arguments originally passed into ``public_api``.
+
+    Returns
+    -------
+    Result from calling `implementation()` or an `__torch_function__`
+    method, as appropriate.
+
+    Raises
+    ------
+    TypeError : if no implementation is found.
+
+    """
+    # Check for __torch_function__ methods.
+    types, overloaded_args = _get_overloaded_types_and_args(relevant_args)
+    # Short-cut for common cases: no overload or only Tensor overload
+    # (directly or with subclasses that do not override __torch_function__).
+    if (not overloaded_args or types == _TENSOR_ONLY):
+        return implementation(*args, **kwargs)
+
+    # Call overrides
+    for overloaded_arg in overloaded_args:
+        # Use `public_api` instead of `implemenation` so __torch_function__
+        # implementations can do equality/identity comparisons.
+        result = overloaded_arg.__torch_function__(
+            public_api, args, kwargs)
+
+        if result is not NotImplemented:
+            return result
+
+    func_name = '{}.{}'.format(public_api.__module__, public_api.__name__)
+    raise TypeError("no implementation found for '{}' on types that implement "
+                    '__torch_function__: {}'
+                    .format(func_name, list(map(type, overloaded_args))))
+
+
+def _verify_matching_signatures(implementation, dispatcher):
+    """Verify that a dispatcher function has the right signature."""
+    implementation_spec = getargspec(implementation)
+    dispatcher_spec = getargspec(dispatcher)
+
+    if (implementation_spec.args != dispatcher_spec.args or
+        implementation_spec.varargs != dispatcher_spec.varargs or
+        implementation_spec.keywords != dispatcher_spec.keywords or
+        (bool(implementation_spec.defaults) !=
+         bool(dispatcher_spec.defaults)) or
+        (implementation_spec.defaults is not None and
+         len(implementation_spec.defaults) !=
+         len(dispatcher_spec.defaults))):
+        raise RuntimeError('implementation and dispatcher for %s have '
+                           'different function signatures' % implementation)
+
+
+_wrapped_func_source = textwrap.dedent("""
+    @functools.wraps(implementation)
+    def {name}(*args, **kwargs):
+        relevant_args = dispatcher(*args, **kwargs)
```

Comment:
(No action needed) Thinking out loud: since we're literally generating Python code to generate these wrappers, I wonder if we can do some sort of partial evaluation so that we can inline the dispatcher here, so we're literally just doing a test to see if arguments are tensors before actually calling the underlying or going the slowpath. Might speed things up.

Reply:
(If we did it this way, you probably wouldn't define the dispatcher as actual functions. Just some data.)